FirebaseCore/Sources/FIRApp.m

```diff
@@ -339,7 +345,7 @@ - (instancetype)initInstanceWithName:(NSString *)name options:(FIROptions *)opti
     _options.editingLocked = YES;
     _isDefaultApp = [name isEqualToString:kFIRDefaultAppName];
     _container = [[FIRComponentContainer alloc] initWithApp:self];
-    _heartbeatLogger = [[FIRHeartbeatLogger alloc] initWithAppID:self.options.googleAppID];
+    _heartbeatLogger = [[self class] heartbeatLoggerProviderForAppID](options.googleAppID);
```

Comment:
Since clients don't directly initialize a FIRApp, I need to find another way to be able to statically inject a fake heartbeat logger.

My current idea is to create a static property `heartbeatLoggerProviderForAppID` that is a block that is called by the internal implementation (line above) to return an initialized heartbeat logger. This block could then be overridden or set by the unit tests to inject a fake heartbeat logger.
```objc
[[self class] heartbeatLoggerProviderForAppID](options.googleAppID)
```
does look a bit wonky though.

Another (similar and maybe more readable?) approach would be to swap the heartbeatLoggerProvider block for a new factory type (`FIRHeartbeatLoggerFactory`) that conforms to a `FIRHeartbeatLoggerFactoryProtocol`. this line would then turn into something like:
```objc
_heartbeatLogger = 
      [[[self class] heartbeatLoggerFactory] makeHeartbeatLoggerWithAppID:options.googleAppID];
```
Similarly, a fake `heartbeatLoggerFactory` could be statially injected during tests to return a predefined fake heartbeat logger.

Reply:
Hmm, I will experiment with second approach.

Reply:
Discussed approach with Ryan offline.